larpandoracontent/LArHelpers/LArMuonLeadingHelper.cc

```diff
@@ -0,0 +1,524 @@
+
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArMuonLeadingHelper.h"
+
+#include "Pandora/PdgTable.h"
+
+#include "Objects/ParticleFlowObject.h"
+#include "Objects/CaloHit.h"
+
+namespace lar_content
+{
+
+using namespace pandora;
+
+LArMuonLeadingHelper::ValidationParameters::ValidationParameters() :
+    LArMCParticleHelper::PrimaryParameters(),
+    m_maxBremsstrahlungSeparation(2.5f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------    
+
+bool LArMuonLeadingHelper::IsDeltaRay(const MCParticle *const pMCParticle)
+{
+    const MCParticleList parentList(pMCParticle->GetParentList());
+
+    if(parentList.empty())
+        return false;
+
+    if (parentList.size() != 1)
+        throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
+
+    if (!LArMCParticleHelper::IsCosmicRay(parentList.front()))
+        return false;
+
+    const LArMCParticle *const pLArMCParticle(dynamic_cast<const LArMCParticle*>(pMCParticle));
+    
+    return (pLArMCParticle->GetProcess() == MC_PROC_MU_IONI);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------    
+
+bool LArMuonLeadingHelper::IsMichel(const MCParticle *const pMCParticle)
+{
+    const MCParticleList parentList(pMCParticle->GetParentList());
+
+    if(parentList.empty())
+        return false;
+    
+    if (parentList.size() != 1)
+        throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
+    
+    if (!LArMCParticleHelper::IsCosmicRay(parentList.front()))
+        return false;    
+    
+    const LArMCParticle *const pLArMCParticle(dynamic_cast<const LArMCParticle*>(pMCParticle));
+
+    return (pLArMCParticle->GetProcess() == MC_PROC_DECAY);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------    
+
+bool LArMuonLeadingHelper::IsMuonLeading(const MCParticle *const pMCParticle)
+{
+    const MCParticle *const pParentMCParticle(LArMCParticleHelper::GetParentMCParticle(pMCParticle));
+    
+    return ((LArMCParticleHelper::GetHierarchyTier(pMCParticle) == 1) && (LArMCParticleHelper::IsCosmicRay(pParentMCParticle)));
+}     
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+const MCParticle *LArMuonLeadingHelper::GetLeadingParticle(const MCParticle *const pMCParticle)
+{
+    MCParticleVector mcParticleVector;
+
+    const MCParticle *pParentMCParticle = pMCParticle;
+    mcParticleVector.push_back(pParentMCParticle);
```

Comment:
Do we need this vector? I think if you get beyond the if statement in the first loop iteration, you're guaranteed to find a leading particle to assign to pParentMCParticle, so maybe you can just set a found flag and simply return the pointer if found, throw an exception if not?

Reply:
have changed this to:

`const MCParticle *LArMuonLeadingHelper::GetLeadingParticle(const MCParticle *const pMCParticle)
{
    if (!pMCParticle)
        throw StatusCodeException(STATUS_CODE_FAILURE);

    if (LArMCParticleHelper::GetHierarchyTier(pMCParticle) == 0)
        throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
        
    const MCParticle *pParentMCParticle = pMCParticle;

    while (LArMCParticleHelper::GetHierarchyTier(pParentMCParticle) != 1))
    {
        const MCParticleList &parentList(pParentMCParticle->GetParentList());
        
        if (parentList.size() != 1)
            throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);

        pParentMCParticle = parentList.front();
    }

    return pParentMCParticle;
} `

Reply:
the idea being to find the address of the MCParticle below the cosmic ray muon in the hierarchy